### PR TITLE
[mlir][Interfaces][NFC] `ValueBoundsConstraintSet`: Pass stop condition in the constructor

### PR DESCRIPTION
**Pull request by matthias-springer**
This commit changes the API of `ValueBoundsConstraintSet`: the stop condition is now passed to the constructor instead of `processWorklist`. That makes it easier to add items to the worklist multiple times and process them in a consistent manner. The current `ValueBoundsConstraintSet` is passed as a reference to the stop function, so that the stop function can be defined before the the `ValueBoundsConstraintSet` is constructed.

This change is in preparation of adding support for branches.

### CODE REVIEW

File: mlir/include/mlir/Interfaces/ValueBoundsOpInterface.h [hunk 1]
```diff
@@ -316,6 +317,9 @@ class ValueBoundsConstraintSet {
 
   /// Builder for constructing affine expressions.
   Builder builder;
+
+  /// The current stop condition function.
+  StopConditionFn stopCondition = nullptr;
```
Comment (MacDue) [inline mlir/include/mlir/Interfaces/ValueBoundsOpInterface.h:322]: Just wondering if this should be a `std::function` instead? `function_ref` begin non-owning could lead to some surprises. E.g. by doing `ValueBoundsConstraintSet cstr(..., /*stopCondition=*/[&]{ ... })`.

Comment (matthias-springer) [reply mlir/include/mlir/Interfaces/ValueBoundsOpInterface.h:322]: Actually I ran into this exact issue in the past (in a different part of the codebase) and it was difficult to debug. Changed it to `std::function`.

File: mlir/include/mlir/Interfaces/ValueBoundsOpInterface.h [hunk 2]
```diff
@@ -113,8 +113,9 @@ class ValueBoundsConstraintSet {
   ///
   /// The first parameter of the function is the shaped value/index-typed
   /// value. The second parameter is the dimension in case of a shaped value.
-  using StopConditionFn =
-      function_ref<bool(Value, std::optional<int64_t> /*dim*/)>;
+  /// The third parameter is this constraint set.
+  using StopConditionFn = std::function<bool(
+      Value, std::optional<int64_t> /*dim*/, ValueBoundsConstraintSet &cstr)>;
```
Comment (kuhar) [inline mlir/include/mlir/Interfaces/ValueBoundsOpInterface.h:118]: This comment is not directly related to the changes in this PR, but It's not immediately clear to me if which value indicates stop/continuation. I like the `walk` function much more which exposes `WalkResult::interrup()` and `WalkRedult::advance()`, but I think that changing to `LogicalResult` and renaming this could also help.
Would be cool if we could revisit this separately.

**Review by MacDue — APPROVED** [review]

File: mlir/include/mlir/Interfaces/ValueBoundsOpInterface.h (hunk 2 above)
Comment (matthias-springer) [reply mlir/include/mlir/Interfaces/ValueBoundsOpInterface.h:122]: Good point, I am also thinking of splitting up `stopCondition` into two functions, so that IR traversal boundary and values that are projected out can be specified separately. We had a use case for that recently.

Comment (matthias-springer) [reply mlir/include/mlir/Interfaces/ValueBoundsOpInterface.h:122]: I'm going to prepare a follow-up PR for this, after my chain of PRs has landed.
